File: leetcode/111-minimum-depth-of-binary-tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
struct Node {
    TreeNode* node;
    int level;
    Node(): node(nullptr), level(0) {}
    Node(TreeNode *node, int level): node(node), level(level) {}
};
class Solution {
public:
    int minDepth(TreeNode *root) {
        queue<Node> nodes;
        if (root != nullptr) {
            nodes.push(Node(root, 1));
        }
        while (!nodes.empty()) {
            Node tmp(nodes.front());
            nodes.pop();
            if (tmp.node->left != nullptr) {
                nodes.push(Node(tmp.node->left, tmp.level + 1));
            }
            if (tmp.node->right != nullptr) {
                nodes.push(Node(tmp.node->right, tmp.level + 1));
            }
            if (tmp.node->left == nullptr && tmp.node->right == nullptr) {
                return tmp.level;
            }
        }
        return 0;
    }
};

File: leetcode/111-minimum-depth-of-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def min_depth(root)
  q = [root] 
  l = 0
  while true
    level = q.select {|i| not i.nil?}
    break if level.empty?
    l += 1
    leaf_idx = level.index {|i| i.left.nil? and i.right.nil?}
    break unless leaf_idx.nil?
    q = level.flat_map {|x| [x.left, x.right]}
  end
  l
end
